packages/material-ui/src/Portal/Portal.js

```diff
@@ -28,14 +28,14 @@ const Portal = React.forwardRef(function Portal(props, ref) {
     }
   }, [container, disablePortal]);
 
+  React.useImperativeHandle(ref, () => mountNode || childRef.current, [mountNode]);
+
   useEnhancedEffect(() => {
     if (onRendered && mountNode) {
       onRendered();
     }
   }, [mountNode, onRendered]);
 
-  React.useImperativeHandle(ref, () => mountNode || childRef.current, [mountNode]);
```

Comment:
Change the order so `ref` triggers before `onRendered`.

Reply:
Wait refs should never trigger before layout effects. Those are two different phases. Do you have a test for that?

Reply:
`useImperativeHandle` triggers at the same time than synchronous layout effects from what I can log (not layout effects!). I'm adding a test case.

Reply:
Huh, need to play around with this. Did not know that.

Reply:
I believe that reason is the following: you need to have access to the DOM element in the synchronous layout effects. If the ref triggers after, you don't have it.

Reply:
I meant react internals. Refs and layout effects are two different phases (like render phase and commit phase) and I thought the commit phase only started after all refs are set. This means that even if you have layout, ref, layout in your render method you always get ref, layout, layout just as if you have layout, passive, layout you get layout, layout, passive.

So yeah I mispoke earlier:
```diff
- Wait refs should never trigger before layout effects.
+ Wait refs should always trigger before layout effects.
```
Just need to rebase this convo :smile:

Reply:
Yeah seems like useImperativeHandle runs not in the ref phase but commit phase: https://codesandbox.io/s/timings-of-imperative-layout-and-passive-effect-ogo9f

Going to open an issue in the react repo since this means that you get inconsistent behavior if a forwarded ref is passed to a host component or useImperativeHandle.

Reply:
Did some experimenting and it is essentially https://github.com/facebook/react/issues/7769 i.e. within a  render block useImperativeHandle runs in the same phase as layout effects but any layout effect queued outside (e.g. in a parent) will run after useImperativeHandle of children.

Reply:
I think `onRendered` is actually an Anti-Pattern. We already have callback refs to handle when a host component is committed. What should happen is

```diff
- const portalRef = React.useRef();
- <Portal container={container} onRendered={() => doSomethingWith(portalRef.current)} ref={portalRef} />
+ <Portal container={container} ref={doSomethingWith} />

Reply:
Oh, interesting. I have never thought about it

Reply:
I think we can re-evaluate dropping onRendered for callback refs in v5. For now this fix works.

Reply:
I'm in favor of dropping the prop in v5. I will update the Modal implementation to stop using it in the next Modal pull request. Thanks for raising. I agree that it doesn't provide anything since the forwardRef effort was completed. 👌